File: src/concurrency-code/polyglotSTM/java/clojure/Account.java
/***
 * Excerpted from "Programming Concurrency on the JVM",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material, 
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose. 
 * Visit http://www.pragmaticprogrammer.com/titles/vspcon for more book information.
***/
package com.agiledeveloper.pcj;

import clojure.lang.Ref;
import clojure.lang.LockingTransaction;
import java.util.concurrent.Callable;

public class Account {
  final private Ref balance;

  public Account(final int initialBalance) throws Exception {
    balance = new Ref(initialBalance);
  }

  public int getBalance() { return (Integer) balance.deref(); }

  public void deposit(final int amount) throws Exception {
    LockingTransaction.runInTransaction(new Callable<Boolean>() {
      public Boolean call()  {
        if(amount > 0) {
          final int currentBalance = (Integer) balance.deref();
          balance.set(currentBalance + amount);
          System.out.println("deposit " + amount + "... will it stay");
          return true;
        } else throw new RuntimeException("Operation invalid");
      }
    });
  }

  public void withdraw(final int amount) throws Exception {
    LockingTransaction.runInTransaction(new Callable<Boolean>() {
      public Boolean call() {
        final int currentBalance = (Integer) balance.deref();
        if(amount > 0 && currentBalance >= amount) {
          balance.set(currentBalance - amount);
          return true;
        } else throw new RuntimeException("Operation invalid");
      }
    });
  }
}

File: bin/concurrency-code/polyglotSTM/jruby/clojure/account.rb
#---
# Excerpted from "Programming Concurrency on the JVM",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material, 
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose. 
# Visit http://www.pragmaticprogrammer.com/titles/vspcon for more book information.
#---
require 'java'
java_import 'clojure.lang.Ref'
java_import 'clojure.lang.LockingTransaction'

class Account
  def initialize(initialBalance)
    @balance = Ref.new(initialBalance)
  end
  
  def balance
    @balance.deref
  end  
  
  def deposit(amount)
    LockingTransaction.run_in_transaction do
      if amount > 0
        @balance.set(@balance.deref + amount)
        puts "deposited $#{amount}... will it stay"
      else
        raise "Operation invalid"
      end
    end
  end

  def withdraw(amount)
    LockingTransaction.run_in_transaction do
      if amount > 0 && @balance.deref >= amount
        @balance.set(@balance.deref - amount)
      else
        raise "Operation invalid"
      end
    end
  end
end

def transfer(from, to, amount)
  LockingTransaction.run_in_transaction do
    to.deposit(amount)
    from.withdraw(amount)
  end
end

def transfer_and_print(from, to, amount)
  begin
    transfer(from, to, amount)
  rescue => ex
    puts "transfer failed #{ex}"
  end

  puts "Balance of from account is #{from.balance}"
  puts "Balance of to account is #{to.balance}"
end

account1 = Account.new(2000)
account2 = Account.new(100)

transfer_and_print(account1, account2, 500)
transfer_and_print(account1, account2, 5000)
